sumcheck/src/verifier/mod.rs

```diff
@@ -0,0 +1,155 @@
+// Copyright (c) Facebook, Inc. and its affiliates.
+//
+// This source code is licensed under the MIT license found in the
+// LICENSE file in the root directory of this source tree.
+
+use alloc::vec::Vec;
+
+use air::LogUpGkrEvaluator;
+use crypto::{ElementHasher, RandomCoin};
+use math::FieldElement;
+
+use crate::{
+    evaluate_composition_poly, EqFunction, FinalLayerProof, FinalOpeningClaim, RoundProof,
+    SumCheckProof, SumCheckRoundClaim,
+};
+
+/// Verifies a round of the sum-check protocol without executing the final check.
+pub fn verify_rounds<E, C, H>(
+    claim: E,
+    round_proofs: &[RoundProof<E>],
+    coin: &mut C,
+) -> Result<SumCheckRoundClaim<E>, SumCheckVerifierError>
+where
+    E: FieldElement,
+    C: RandomCoin<Hasher = H, BaseField = E::BaseField>,
+    H: ElementHasher<BaseField = E::BaseField>,
+{
+    let mut round_claim = claim;
+    let mut evaluation_point = vec![];
+    for round_proof in round_proofs {
+        let round_poly_coefs = round_proof.round_poly_coefs.clone();
+        coin.reseed(H::hash_elements(&round_poly_coefs.0));
+
+        let r = coin.draw().map_err(|_| SumCheckVerifierError::FailedToGenerateChallenge)?;
+
+        round_claim = round_proof.round_poly_coefs.evaluate_using_claim(&round_claim, &r);
+        evaluation_point.push(r);
+    }
+
+    Ok(SumCheckRoundClaim {
+        eval_point: evaluation_point,
+        claim: round_claim,
+    })
+}
+
+/// Verifies sum-check proofs, as part of the GKR proof, for all GKR layers except for the last one
+/// i.e., the circuit input layer.
+pub fn verify_sum_check_intermediate_layers<
+    E: FieldElement,
+    C: RandomCoin<Hasher = H, BaseField = E::BaseField>,
+    H: ElementHasher<BaseField = E::BaseField>,
+>(
+    proof: &SumCheckProof<E>,
+    gkr_eval_point: &[E],
+    claim: (E, E),
+    transcript: &mut C,
+) -> Result<FinalOpeningClaim<E>, SumCheckVerifierError> {
+    // generate challenge to batch sum-checks
+    transcript.reseed(H::hash_elements(&[claim.0, claim.1]));
+    let r_batch: E = transcript
+        .draw()
+        .map_err(|_| SumCheckVerifierError::FailedToGenerateChallenge)?;
+
+    // compute the claim for the batched sum-check
+    let reduced_claim = claim.0 + claim.1 * r_batch;
+
+    let SumCheckProof { openings_claim, round_proofs } = proof;
+
+    let final_round_claim = verify_rounds(reduced_claim, round_proofs, transcript)?;
+    assert_eq!(openings_claim.eval_point, final_round_claim.eval_point);
+
+    let p0 = openings_claim.openings[0];
+    let p1 = openings_claim.openings[1];
+    let q0 = openings_claim.openings[2];
+    let q1 = openings_claim.openings[3];
+
+    let eq = EqFunction::new(gkr_eval_point.to_vec()).evaluate(&openings_claim.eval_point);
+
+    if (p0 * q1 + p1 * q0 + r_batch * q0 * q1) * eq != final_round_claim.claim {
```

Comment:
I think this is the same equation as computed by `comb_func()`, right? If so, should we use it here?

Reply:
Correct!
Changed it now